scripts/ccpp_datafile.py

```diff
@@ -288,11 +400,72 @@ def _retrieve_dependencies(table):
     # end for
     return sorted(result)
 
+###############################################################################
+def _retrieve_dyn_const_routines(table):
+###############################################################################
+    """Find and return a list of all scheme constituent routines.
+    # Test valid dynamic constituent routines
+    >>> table = ET.fromstring("<ccpp_datatable version='1.0'><dyn_const_routines>" \
+               "<dyn_const_routine parent='banana'>dyn_const_get"    \
+               "</dyn_const_routine><dyn_const_routine>dyn_const_2"  \
+               "</dyn_const_routine></dyn_const_routines></ccpp_datatable>")
+    >>> _retrieve_dyn_const_routines(table)
+    ['dyn_const_2', 'dyn_const_get']
+
+    # Test no dynamic constituent routines
+    >>> table = ET.fromstring("<ccpp_datatable version='1.0'><dyn_const_routines>" \
+                              "</dyn_const_routines></ccpp_datatable>")
+    >>> _retrieve_dyn_const_routines(table)
+    []
+
+    # Test missing dynamic constituent routines tag
+    >>> table = ET.fromstring("<ccpp_datatable version='1.0'></ccpp_datatable>")
+    >>> _retrieve_dyn_const_routines(table)
+    Traceback (most recent call last):
+    ...
+    ccpp_datafile.CCPPDatatableError: Could not find 'dyn_const_routines' element
+
+    """
+    result = set()
+    routines = table.find("dyn_const_routines")
+    if routines is None:
+        raise CCPPDatatableError("Could not find 'dyn_const_routines' element")
+    # end if
+    for routine in routines:
+        routine_name = routine.text
+        if routine_name is not None:
+            result.add(routine_name)
+        # end if
+    # end for
+    return sorted(result)
```

Comment:
```suggestion
    routine_names = [routine.text for routine in routines if routine.text]
    return sorted(routine_names)
```
_Typically_, list comprehensions are more optimized for generating lists but if the current code is more readable for everyone, please disregard!

Also, if we want to add the `routine_name` even if it is an empty string, this is perfectly fine.  But if we want to exclude empty strings, we might want to change this to the check in the list comprehension `if routine.text` as empty strings are considered `False`.

Reply:
changed!